wxAUI notebook pages dragging-related fixes

vadz opened the pull request:
This fixes user-visible #24890 and #24892, but, as discussed in the latter, I'm not sure if we shouldn't restore the old behaviour of `ShowHint()` and pass it the rectangle in screen coordinates.

vadz committed:
Use correct DPI for wxTabFrame used by wxAuiNotebook in wxMSW

This pseudo-window doesn't have a valid HWND, so the base class
implementation of GetDPI() doesn't work for it and falls back to the
primary monitor DPI, meaning that it returned incorrect results if the
notebook was on a monitor using different DPI.

Fix this by forwarding GetDPI() to wxAuiTabCtrl which is a real window
and so uses the appropriate DPI.

vadz committed:
Simplify code initializing pane best size in wxAuiManager

No real changes, just use wxSize::IncTo() instead of writing it out
manually.

vadz committed:
Use current, and not best, size when adding panes to wxAuiManager

This notably ensures that a pane being drag-and-dropped keeps its
existing size after being dropped in a new location instead of shrinking
to its best size, which might be not set at all, resulting in pane
becoming effectively invisible.

Closes #24890.

vadz committed:
Allow setting flags of wxAuiManager used by wxAuiNotebook

Add wxAuiNotebook::SetManagerFlags() and use it in the sample to allow
testing how the different flags, such as wxAUI_MGR_HINT_FADE, affect
wxAuiNotebook behaviour.

vadz committed:
Fix fading in hints shown when dragging wxAuiNotebook pages

ShowHint() shouldn't be called directly, as it doesn't set m_hintFadeAmt
correctly and so the hint window is always shown directly, even when
wxAUI_MGR_HINT_FADE flag is set (which is the default).

Fix this by adding wxAuiManager::UpdateHint() which does update
m_hintFadeAmt and use it in wxAuiNotebook.

Also document that ShowHint() shouldn't be called directly.

vadz committed:
Default "offset" parameter of CalculateHintRect() to zero

No real changes, just get rid of dummy "zero" variables and don't pass
the offset to CalculateHintRect(), nor DrawHintRect(), at all when it's
zero.

AliKet commented:
[ef2510c](https://github.com/wxWidgets/wxWidgets/pull/24911/commits/ef2510ccea1c6d0084104652e4189aab36bc52c2) shouldn't be included (see https://github.com/wxWidgets/wxWidgets/issues/24892#issuecomment-2439712989)

AliKet commented:
I'm just wondering if it's a good idea to break the `ShowHint()` API in **wx3.3** and make the rectangle it accepts in client coordinates instead of screen coordinates and update the aui code accordingly?

vadz committed:
Restore use of screen coordinates in wxAuiManager::ShowHint()

This function can be overridden in the application code, so it is not a
good idea to silently change its behaviour and it used to the rectangle
in screen coordinates until the changes of 29c8bfc249 (Use wxOverlay to
show docking hint instead of transparent wxFrame, 2024-01-31) when it
was changed to the rectangle in client coordinates of the window
associated with wxAuiManager.

Restore the old behaviour to avoid breaking existing code.

See #24193.

vadz commented:
Thanks, I've updated the PR to revert the change in `ShowHint()`.

I don't think it's worth changing its semantics and I even kept it for the new `UpdateHint()` too because it would be confusing to use different coordinates for the two closely related functions.

AliKet commented on src/aui/auibook.cpp:
```diff
@@ -2712,7 +2712,7 @@ void wxAuiNotebook::OnTabDragMotion(wxAuiNotebookEvent& evt)
                     tab_ctrl->ClientToScreen(&hint_rect.x, &hint_rect.y);
-                    m_mgr.ShowHint(hint_rect);
+                    m_mgr.UpdateHint(tab_ctrl->GetRect());
                     return;
```
I don't understand this change:
- `hint_rect` is calculated but not used!
- We pass a rectangle (in client coordinates) to `UpdateHint()`, but it accepts a rectangle in screen coordinates? or, the intention here is to use `tab_ctrl->GetScreenRect()` instead?

vadz replied:
Oops, thanks for noticing this. This is just a bad rebase conflict resolution, it should be `UpdateHint(hint_rect)`, of course.

I thought about changing the code to use `GetScreenRect()` too, but decided it wasn't worth doing it, but maybe it is...

vadz committed:
Simplify wxAuiNotebook code by using GetScreenRect()

This is slightly simpler and more clear than using GetRect() followed by
ClientToScreen().

AliKet commented:
Please, let me know if I should remove this [line](https://github.com/wxWidgets/wxWidgets/blob/b1442f2d4c5554ee587eb9a707caa95716b8ba94/src/aui/framemanager.cpp#L4494) in PR #24921 because it would be incorrect for `wxQt` when applying the aforementioned PR. Or, would it be much more sense to remove it here?

TIA!

vadz commented:
@AliKet I don't understand why do we need this line, so it should probably be removed here, but how exactly can I test this code, i.e. what should be done to see the current behaviour?

AliKet commented:
> @AliKet I don't understand why do we need this line, so it should probably be removed here, but how exactly can I test this code, i.e. what should be done to see the current behaviour?

Just disable **Live Resize Update** under **Options** and try to resize some panes:

![Screenshot from 2024-10-28 17-48-24](https://github.com/user-attachments/assets/8606570a-9ada-46c4-ae97-dd6a98f96f84)

With this PR and #24921 the resize hint under `wxQt` is drawn correctly at the right offset.

vadz commented:
No, they actually don't... Well, maybe wxGTK does (I think it always uses origin of (0,0)) but if you apply this:

```diff
iff --git a/samples/aui/auidemo.cpp b/samples/aui/auidemo.cpp
index 96a51c4ba0..48c6ed62cf 100644
--- a/samples/aui/auidemo.cpp
+++ b/samples/aui/auidemo.cpp
@@ -695,6 +695,10 @@ MyFrame::MyFrame(wxWindow* parent,
                  long style)
         : wxFrame(parent, id, title, pos, size, style)
 {
+    auto tb = CreateToolBar(wxTB_TEXT);
+    tb->AddTool(wxID_ABOUT, "About", wxIcon(sample_xpm));
+    tb->Realize();
+
     // tell wxAuiManager to manage this frame
     m_mgr.SetManagedWindow(this);
 
```

to make the origin non-0 in wxMSW (admittedly, it's uncommon to use normal toolbar in a program also using wxAUI, but it's definitely possible), you can see that the position is wrong there too, **both** with and without that line.

To make things work correctly in wxMSW you actually need to use

```cpp
                rect.SetPosition(rect.GetPosition() - m_frame->GetClientAreaOrigin());
```

and now wxQt behaviour makes even less sense to me, how could it have needed `+` here?

AliKet commented:
> 
> To make things work correctly in wxMSW you actually need to use
> 
> ```c++
>                 rect.SetPosition(rect.GetPosition() - m_frame->GetClientAreaOrigin());
> ```
> 
> and now wxQt behaviour makes even less sense to me, how could it have needed `+` here?

For `wxQt` I missed a subtlety when I made this PR which should have looked like this:

```
diff --git a/src/qt/dcclient.cpp b/src/qt/dcclient.cpp
index 46247c09a1..675d127b20 100644
--- a/src/qt/dcclient.cpp
+++ b/src/qt/dcclient.cpp
@@ -115,10 +115,10 @@ wxClientDCImpl::wxClientDCImpl( wxDC *owner, wxWindow *win )
 {
      if ( m_ok )
      {
-        m_qtPainter->translate( wxQtConvertPoint(win->GetClientAreaOrigin()) );
-
         m_qtPainter->setClipRect( wxQtConvertRect(win->GetClientRect()),
                                   m_clipping ? Qt::IntersectClip : Qt::ReplaceClip );
+
+        m_qtPainter->translate( wxQtConvertPoint(win->GetClientAreaOrigin()) );
     }
 }
 
```

So apparently this line
```c++
                 rect.SetPosition(rect.GetPosition() - m_frame->GetClientAreaOrigin());
```
 is only needed under wxMSW (wxOSX?)

vadz commented:
Ugh, something is very wrong here. First of all, thinking more about this, using `-` for wxMSW really doesn't make sense: either the offset is accounted for by `wxClientDC` and `wxOverlayDC` based on it and then this line shouldn't be used at all or it isn't and in this case `+` would be correct to compensate for it. Second, somehow using `-` does work correctly when starting the sample but if you undock the AUI toolbars (which do not count to the client area offset, AFAICS), it doesn't work correctly for the splitter between the tree and the notebook any — while it still does for the splitter between them both and the text at the bottom.

So it looks like action part position itself is incorrect, maybe?

But in any case, the first question to ask is what is the point (0,0) of `wxOverlayDC` associated with the window. IMO it should be the same as the point (0,0) of `wxClientDC` and hence must be just below the toolbar, i.e. at `GetClientAreaOrigin()` in "real" coordinates. Is this the case for wxQt?

vadz commented:
FWIW here is a test showing that currently `wxOverlay` origin is broken under MSW:

```cpp
#include <wx/wx.h>
#include <wx/overlay.h>

class MyFrame : public wxFrame
{
public:
    MyFrame() : wxFrame(nullptr, wxID_ANY, "Test")
    {
        auto tb = CreateToolBar();
        tb->AddTool(wxID_ABOUT, "About", wxICON(sample));
        tb->Realize();

        Show();

        wxOverlayDC dc(m_overlay, this);
        dc.Clear();
        dc.SetBrush(*wxRED_BRUSH);
        dc.SetPen(*wxTRANSPARENT_PEN);
        dc.DrawRectangle(0, 0, 200, 100);
    }

    wxOverlay m_overlay;
};

class MyApp : public wxApp
{
public:
    bool OnInit() override
    {
        new MyFrame();
        return true;
    }
};
wxIMPLEMENT_APP(MyApp);
```

If you run this, you can see that the overlay is offset from the toolbar by toolbar height.

AliKet commented:
> But in any case, the first question to ask is what is the point (0,0) of `wxOverlayDC` associated with the window. IMO it should be the same as the point (0,0) of `wxClientDC` and hence must be just below the toolbar, i.e. at `GetClientAreaOrigin()` in "real" coordinates. Is this the case for wxQt?

Yes and here is how it looks like if we give it a cyan colour (with the corrected patch I mentioned above applied):

![Screenshot from 2024-10-28 21-44-42](https://github.com/user-attachments/assets/4d670201-0ade-4a51-9b3a-9b8fc9c7ef5e)

wxGTK3 using Wayland is exactly the same.

AliKet commented:
> FWIW here is a test showing that currently `wxOverlay` origin is broken under MSW:
> 
> ```c++
> #include <wx/wx.h>
> #include <wx/overlay.h>
> 
> class MyFrame : public wxFrame
> {
> public:
>     MyFrame() : wxFrame(nullptr, wxID_ANY, "Test")
>     {
>         auto tb = CreateToolBar();
>         tb->AddTool(wxID_ABOUT, "About", wxICON(sample));
>         tb->Realize();
> 
>         Show();
> 
>         wxOverlayDC dc(m_overlay, this);
>         dc.Clear();
>         dc.SetBrush(*wxRED_BRUSH);
>         dc.SetPen(*wxTRANSPARENT_PEN);
>         dc.DrawRectangle(0, 0, 200, 100);
>     }
> 
>     wxOverlay m_overlay;
> };
> 
> class MyApp : public wxApp
> {
> public:
>     bool OnInit() override
>     {
>         new MyFrame();
>         return true;
>     }
> };
> wxIMPLEMENT_APP(MyApp);
> ```
> 
> If you run this, you can see that the overlay is offset from the toolbar by toolbar height.

Thanks for the code, I'll try to take a look at it tomorrow as I'm exhausted right now.

Regards.

vadz commented:
> I'll try to take a look at it tomorrow as I'm exhausted right now.

I'll look at wxMSW myself, something seems very wrong there as `wxClientDC` is not offset correctly neither...

MaartenBent commented:
Maybe you already found the problem, but I was also intrigued. I think the problem is `GetClientRect`/`GetClientAreaOrigin` adds the `wxToolbar` height (or width) to the origin, but when converting client coordinates to screen coordinates this is not accounted for.
And Windows doesn't know that the origin is shifted, so the screen coordinates will be shifted as well.

vadz commented:
Thanks, there is indeed this, but also the fact that the adjustment for the client area origin in `wxClientDCImpl::InitDC()` is not done for wxMSW, while IMO it should be.

This won't be difficult to fix but it's strange that this hasn't been noticed until now, I spent some time looking for the changes that could have broken it but it really looks like it always worked like this. I guess people don't use `wx{Client,Paint}DC` with frames with toolbars that often, but still...

vadz commented:
Just to be clear (as much for my own benefit as for anyone else): fixing `wxOverlay` right now is as simple as doing this:

```diff
diff --git a/src/msw/overlay.cpp b/src/msw/overlay.cpp
index 9d8240efa3..449a906d29 100644
--- a/src/msw/overlay.cpp
+++ b/src/msw/overlay.cpp
@@ -161,7 +161,7 @@ void wxOverlayImpl::Init(wxDC* dc, int , int , int , int )
     m_window = dc->GetWindow();
 
     // The rectangle must be in screen coordinates
-    m_rect = m_window->GetClientRect();
+    m_rect = m_window->GetClientSize();
     m_window->ClientToScreen(&m_rect.x, &m_rect.y);
 
     if ( IsUsingConstantOpacity() )
```

but if you replace `wxOverlayDC dc(m_overlay, this)` in the example above with `wxClientDC dc(this)` you can see that it's also wrong and to fix it, and `wxPaintDC`, we need

```diff
diff --git a/src/msw/dcclient.cpp b/src/msw/dcclient.cpp
index cc2549ca7c..5818562618 100644
--- a/src/msw/dcclient.cpp
+++ b/src/msw/dcclient.cpp
@@ -188,9 +188,8 @@ void wxClientDCImpl::InitDC()
 {
     wxWindowDCImpl::InitDC();
 
-    // in wxUniv build we must manually do some DC adjustments usually
-    // performed by Windows for us
-#if defined(__WXUNIVERSAL__)
+    // Account for the origin of the client area which is non-zero only for
+    // TLWs with (left or top) toolbar: we shouldn't draw over the toolbar.
     wxPoint ptOrigin = m_window->GetClientAreaOrigin();
     if ( ptOrigin.x || ptOrigin.y )
     {
@@ -198,6 +197,9 @@ void wxClientDCImpl::InitDC()
         SetDeviceOrigin(ptOrigin.x, ptOrigin.y);
     }
 
+    // in wxUniv build we must manually do some DC adjustments usually
+    // performed by Windows for us
+#if defined(__WXUNIVERSAL__)
     // clip the DC to avoid overwriting the non client area
     wxSize size = m_window->GetClientSize();
     DoSetClippingRegion(0, 0, size.x, size.y);
```

but if we do this, then the fix above is not enough for `wxOverlay` any more and we need to also do something like

```cpp
    m_rect.Offset(-m_window->GetClientAreaOrigin());
```

in the overlay code (we can't do it before client-to-screen conversion).

But this actually doesn't seem too bad, and looks like the right thing to do, so I'll make a PR doing this, unless somebody quickly tells me why this is a bad idea.

vadz committed:
Remove unnecessary offset of hint rectangle by client area origin

This shouldn't be necessary and won't be, once the related bugs in the
different ports (see #24924 for wxMSW) are corrected.

vadz committed:
Remove unneeded code and variable in wxAuiManager code

Remove some remnants of the code resetting m_actionHintRect which became
unnecessary many years ago and don't use a temporary variable when we
can just assign to m_actionHintRect itself directly instead.

No real changes.

vadz commented:
I've created the PR promised above now and pushed the commit making things work correctly for me under MSW, even in the presence of a toolbar.

vadz commented:
I'm merging this and the related PRs to fix the user-visible problems, if necessary we can always improve things (e.g. by switching to child overlay windows) later, but for now this seems good enough.
